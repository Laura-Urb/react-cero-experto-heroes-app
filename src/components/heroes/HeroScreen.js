import React, { useMemo } from 'react';
import { Redirect, useParams } from 'react-router-dom';
import { heroImages } from '../../helpers/herosImages';
import { getHeroesById } from '../../selectors/getHeroesById';

export const HeroScreen = ({ history }) => {
  const { heroeId } = useParams();
  const heroe = useMemo(() => getHeroesById(heroeId), [heroeId]);
  if (!heroe) {
    return <Redirect to='/' />;
  }
  const { superhero, alter_ego, first_appearance, characters, publisher } =
    heroe;

  const handleReturn = () => {
    if (history.length <= 2) history.push('/');
    else history.goBack();
  };

  return (
    <div className='row mt-5'>
      <div className='col-4'>
        <img
          // src={`../assets/heroes/${heroeId}.jpg`}
          src={heroImages(`./${heroeId}.jpg`).default}
          className='img-thumbnail animate__animated animate__fadeInLeft'
          alt={superhero}
        />
      </div>
      <div className='col-8'>
        <h3>{superhero}</h3>
        <ul className='list-group mb-3'>
          <li className='list-group-item'>Alter ego: {alter_ego}</li>
          <li className='list-group-item'>Publisher: {publisher}</li>
          <li className='list-group-item'>
            First Appearance: {first_appearance}
          </li>
        </ul>
        <h5>Character</h5>
        <p>{characters}</p>
        {/* <Link to={`./heroe/${id}`}>MÃ¡s..</Link> */}
        <button className='btn btn-outline-info' onClick={handleReturn}>
          Volver
        </button>
      </div>
    </div>
  );
};
