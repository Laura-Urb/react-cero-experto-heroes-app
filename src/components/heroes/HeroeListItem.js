import React from 'react';
import { Link } from 'react-router-dom';
import { heroImages } from '../../helpers/herosImages';

export const HeroeListItem = ({
  id,
  superhero,
  alter_ego,
  first_appearance,
  characters,
}) => {
  return (
    <div className='col animate__animated animate__fadeIn'>
      <div className='card mb-3' style={{ maxWidth: 540 }}>
        <div className='row g-0'>
          <div className='col-md-4'>
            <img
              src={heroImages(`./${id}.jpg`).default}
              className='img-fluid rounded-start'
              alt={superhero}
            />
          </div>
          <div className='col-md-8'>
            <div className='card-body'>
              <h5 className='card-title'>{superhero}</h5>
              <p className='card-text'>{alter_ego}</p>
              {alter_ego !== characters && (
                <p className='card-text'>{characters}</p>
              )}
              <p className='text-muted'>{first_appearance}</p>
              <Link to={`./heroe/${id}`}>MÃ¡s..</Link>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
