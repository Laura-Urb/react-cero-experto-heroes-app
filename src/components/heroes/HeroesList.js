import React, { useMemo } from 'react';
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';
import { HeroeListItem } from './HeroeListItem';

export const HeroesList = ({ publisher }) => {
  const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher]);
  return (
    <div className='row row-cols-1 row-cols-md-3 g-4'>
      {heroes.map((heroe) => (
        <HeroeListItem key={heroe.id} {...heroe} />
      ))}
    </div>
  );
};
