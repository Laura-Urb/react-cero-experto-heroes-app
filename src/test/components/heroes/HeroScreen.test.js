import { mount, shallow } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { HeroScreen } from '../../../components/heroes/HeroScreen';

describe('Prueba HeroScreen', () => {
  const historyMock = {
    push: jest.fn(),
    length: 10,
    goBack: jest.fn(),
  };

  test('debe mostrarse el componente redirect si no hay argumentos en el url ', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/heroe']}>
        <HeroScreen history={historyMock} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Redirect').exists()).toBe(true);
  });

  test('debe mostrarse un heroes si el parameturo existe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
        <Route path='/heroe/:heroeId' component={HeroScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find('.row').exists()).toBe(true);
  });

  test('debe de regresar a la pantalla home con push', () => {
    const historyMock = {
      push: jest.fn(),
      length: 1,
      goBack: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
        <Route
          path='/heroe/:heroeId'
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );
    wrapper.find('button').prop('onClick')();
    expect(historyMock.push).toHaveBeenCalledWith('/');
    expect(historyMock.goBack).not.toHaveBeenCalled();
  });

  test('debe de regresar a la pantalla anterior', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
        <Route
          path='/heroe/:heroeId'
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );
    wrapper.find('button').prop('onClick')();
    expect(historyMock.push).toHaveBeenCalledTimes(0);
    expect(historyMock.goBack).toHaveBeenCalled();
  });
  test('debe de llamar el redirect si el heroes no existe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/heroe/marvel-spider78787878787878']}>
        <Route
          path='/heroe/:heroeId'
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );
    expect(wrapper.text()).toBe('');
  });
});
