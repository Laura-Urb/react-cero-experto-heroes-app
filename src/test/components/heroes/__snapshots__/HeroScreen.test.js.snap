// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba HeroScreen debe mostrarse el componente redirect si no hay argumentos en el url  1`] = `""`;
